### PR TITLE
Add an ASP.NET app that runs in AWS ECS Fargate

### PR DESCRIPTION
**Pull request by joeduffy**
This adds an example of a basic ASP.NET app that runs in AWS ECS
Fargate. This includes the infrastructure required to build and publish
that application as a Docker container image to a private ECR repo,
spin up a new ECS Fargate cluster, run a load balancer listening for
external Internet traffic on port 80, and run services across all
subnets in the default VPC for the target deployment region.

### CODE REVIEW

File: aws-cs-fargate/App/Program.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+
+namespace dotnet_core_tutorial
```
Comment (mikhailshilkov) [inline aws-cs-fargate/App/Program.cs]: Rename the namespace to something .NET-y

File: aws-cs-fargate/App/appsettings.Development.json [hunk 2]
```diff
@@ -0,0 +1,9 @@
+{
+  "Logging": {
+    "LogLevel": {
+      "Default": "Debug",
+      "System": "Information",
+      "Microsoft": "Information"
+    }
+  }
+}
```
Comment (mikhailshilkov) [inline aws-cs-fargate/App/appsettings.Development.json]: Nit, but I would omit this file

File: aws-cs-fargate/Infra/.gitignore [hunk 3]
```diff
@@ -0,0 +1,2 @@
+/bin/
+/obj/
```
Comment (mikhailshilkov) [inline aws-cs-fargate/Infra/.gitignore]: Maybe merge `.gitignore`s in the root folder of the example?

File: aws-cs-fargate/Infra/Program.cs [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    {
+        return Deployment.RunAsync(async () => {
+            // Read back the default VPC and public subnets, which we will use.
+            var vpc = await Ec2.Invokes.GetVpc(new Ec2.GetVpcArgs { Default = true });
+            var subnet = await Ec2.Invokes.GetSubnetIds(new Ec2.GetSubnetIdsArgs { VpcId = vpc.Id });
+
+            // Create a SecurityGroup that permits HTTP ingress and unrestricted egress.
+            var webSg = new Ec2.SecurityGroup("web-sg", new Ec2.SecurityGroupArgs {
+                VpcId = vpc.Id,
+                Egress = new[] {
```
Comment (mikhailshilkov) [inline aws-cs-fargate/Infra/Program.cs]: `new[]` is not required here and down below

File: aws-cs-fargate/Infra/Program.cs [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+using Ecs = Pulumi.Aws.Ecs;
+using Ecr = Pulumi.Aws.Ecr;
+using Elb = Pulumi.Aws.ElasticLoadBalancingV2;
+using Iam = Pulumi.Aws.Iam;
+
+class Program
+{
+    static Task<int> Main()
+    {
+        return Deployment.RunAsync(async () => {
```
Comment (mikhailshilkov) [inline aws-cs-fargate/Infra/Program.cs]: It's common to put `{` on a new line, also in our .NET code

File: aws-cs-fargate/Infra/Program.cs [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+                var creds = await Ecr.Invokes.GetCredentials(new Ecr.GetCredentialsArgs { RegistryId = rid });
+                var credsData = Convert.FromBase64String(creds.AuthorizationToken);
+                return Encoding.UTF8.GetString(credsData).Split(":");
+            });
+            var image = new Docker.Image("app-img", new Docker.ImageArgs {
+                Build = "../App",
+                ImageName = appRepo.RepositoryUrl,
+                Registry = new Docker.ImageRegistry {
+                    Server = appRepo.RepositoryUrl,
+                    Username = appRepoCreds.Apply(creds => creds[0]),
```
Comment (mikhailshilkov) [inline aws-cs-fargate/Infra/Program.cs]: I think you should be able to do `appRepoCreds.Get(0)`

File: aws-cs-fargate/Infra/Program.cs [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+            // Spin up a load balanced service running our container image.
+            var appTask = new Ecs.TaskDefinition("app-task", new Ecs.TaskDefinitionArgs {
+                Family = "fargate-task-definition",
+                Cpu = "256",
+                Memory = "512",
+                NetworkMode = "awsvpc",
+                RequiresCompatibilities = new[] { "FARGATE" },
+                ExecutionRoleArn = taskExecRole.Arn,
+                ContainerDefinitions = image.ImageName.Apply(imgName => @"[{
```
Comment (mikhailshilkov) [inline aws-cs-fargate/Infra/Program.cs]: `imgName` -> `imageName` in .NET

File: aws-cs-fargate/Infra/Program.cs [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+                        TargetGroupArn = webTg.Arn,
+                        ContainerName = "my-app",
+                        ContainerPort = 80,
+                    },
+                },
+            }, new CustomResourceOptions { DependsOn = { webListener } });
+
+            // Export the resulting web address.
+            return new Dictionary<string, object?>{
+                { "url", webLb.DnsName.Apply(url => $"http://{url}") },
```
Comment (mikhailshilkov) [inline aws-cs-fargate/Infra/Program.cs]: Try changing to `Output.Format`

File: aws-cs-fargate/README.md [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+[ASP.NET's](https://docs.microsoft.com/en-us/aspnet/core/getting-started/?view=aspnetcore-3.1) Getting Started
+tutorials. The result is a simple development experience and yet an end result that uses modern, production-ready AWS
+infrastructure. [`./Infra/Program.cs`](./Infra/Program.cs) defines the project's infrastructure.
+
+## Prerequisites
+
+* [Install Pulumi](https://www.pulumi.com/docs/get-started/install/)
+* [Install .NET Core 3](https://dotnet.microsoft.com/download)
+* [Connect Pulumi to AWS](https://www.pulumi.com/docs/intro/cloud-providers/aws/setup/) (if your AWS CLI is
+      configured, this will just work)
```
Comment (mikhailshilkov) [inline aws-cs-fargate/README.md]: Docker is another pre

File: aws-cs-fargate/README.md [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+     ~   ├─ aws:ecs:Service         app-svc             updated     [diff: ~taskDefinition]
+         └─ docker:image:Image      app-img
+
+    Resources:
+        ~ 1 updated
+        +-1 replaced
+        2 changes. 9 unchanged
+   ```
+
+   Notice that `pulumi up` redeploys just the parts of the application/infrastructure that you've edited.
```
Comment (mikhailshilkov) [inline aws-cs-fargate/README.md]: So does the re-publishing of the docker image work?

File: aws-cs-fargate/App/appsettings.Development.json (hunk 2 above)
Comment (joeduffy) [reply aws-cs-fargate/App/appsettings.Development.json]: Ok! Was autogenerated by `dotnet new`, wasn't sure what it is 😄

File: aws-cs-fargate/App/Program.cs (hunk 1 above)
Comment (joeduffy) [reply aws-cs-fargate/App/Program.cs]: Will do, all of this is whatever `dotnet new` spit out, fwiw.

File: aws-cs-fargate/Infra/Program.cs (hunk 5 above)
Comment (joeduffy) [reply aws-cs-fargate/Infra/Program.cs]: Alright. Not quite sure what the style is meant to be, e.g.

```
    LoadBalancers = {
        x
    },
```

or

```
    LoadBalancers =
    {
        x
    },
```

Do we have plans to add style/linter checking?

File: aws-cs-fargate/Infra/Program.cs (hunk 6 above)
Comment (joeduffy) [reply aws-cs-fargate/Infra/Program.cs]: Hmm: `error CS1061: 'Output<string[]>' does not contain a definition for 'Get' and no accessible extension method 'Get' accepting a first argument of type 'Output<string[]>' could be found (are you missing a using directive or an assembly reference?)`

File: aws-cs-fargate/Infra/Program.cs (hunk 8 above)
Comment (joeduffy) [reply aws-cs-fargate/Infra/Program.cs]: Done.

File: aws-cs-fargate/Infra/Program.cs (hunk 7 above)
Comment (joeduffy) [reply aws-cs-fargate/Infra/Program.cs]: Done.

File: aws-cs-fargate/README.md (hunk 9 above)
Comment (joeduffy) [reply aws-cs-fargate/README.md]: Good catch, fixed.

File: aws-cs-fargate/README.md (hunk 10 above)
Comment (joeduffy) [reply aws-cs-fargate/README.md]: It does! I filed an issue for the lack of diffs (https://github.com/pulumi/pulumi-docker/issues/131) which led me astray. The root problem was that I was using the ECR registry URL and _not_ the built image's full name, which is just the ECR registry plus the hash. The hash is required to ensure the ECS service gets restarted upon a new image being pushed.

File: aws-cs-fargate/Infra/Program.cs (hunk 5 above)
Comment (mikhailshilkov) [reply aws-cs-fargate/Infra/Program.cs]: The latter

File: aws-cs-fargate/Infra/Program.cs (hunk 6 above)
Comment (mikhailshilkov) [reply aws-cs-fargate/Infra/Program.cs]: My bad, should be `GetAt` https://github.com/pulumi/pulumi/blob/master/sdk/dotnet/Pulumi/Core/OutputExtensions.cs#L22

File: aws-cs-fargate/Infra/Program.cs (hunk 5 above)
Comment (joeduffy) [reply aws-cs-fargate/Infra/Program.cs]: Oh my, the code is now twice as long! 🤦‍♂  But I shall resist the temptation to comment on style guidelines ... (oh wait, too late)

**Pull request merged**